ergo-p2p/src/peer_spec.rs

```diff
@@ -0,0 +1,117 @@
+//! docuemt
+use std::convert::TryInto;
+use std::io;
+
+use sigma_ser::vlq_encode::VlqEncodingError;
+use sigma_ser::{ScorexParsingError, ScorexSerializable, ScorexSerializeResult};
+
+use crate::peer_addr::PeerAddr;
+use crate::{peer_feature::PeerFeature, protocol_version::ProtocolVersion};
+
+///
+pub struct PeerSpec {
+    agent_name: String,
+    protocol_version: ProtocolVersion,
+    node_name: String,
+    declared_addr: Option<PeerAddr>,
+    features: Vec<PeerFeature>,
+}
+
+impl PeerSpec {
+    /// Tester
+    pub fn new(
+        agent_name: &str,
+        protocol_version: ProtocolVersion,
+        node_name: &str,
+        declared_addr: Option<PeerAddr>,
+        features: Vec<PeerFeature>,
+    ) -> Self {
+        Self {
+            agent_name: agent_name.into(),
+            protocol_version,
+            node_name: node_name.into(),
+            declared_addr,
+            features,
+        }
+    }
+
+    /// local_addr
+    pub fn local_addr(&self) -> Option<PeerAddr> {
+        Some(self.features.iter().find_map(PeerFeature::as_local_addr)?.0)
+    }
+
+    /// reachable
+    pub fn reachable_peer(&self) -> bool {
+        self.addr().is_some()
+    }
+
+    /// address
+    pub fn addr(&self) -> Option<PeerAddr> {
+        self.declared_addr.or_else(|| self.local_addr())
+    }
+}
+
+impl ScorexSerializable for PeerSpec {
+    fn scorex_serialize<W: sigma_ser::vlq_encode::WriteSigmaVlqExt>(
+        &self,
+        w: &mut W,
+    ) -> ScorexSerializeResult {
+        w.put_short_string(&self.agent_name)?;
+        self.protocol_version.scorex_serialize(w)?;
+        w.put_short_string(&self.node_name)?;
+
+        w.put_option(self.declared_addr, &|w: &mut W,
+                                           addr: PeerAddr|
+         -> io::Result<()> {
+            let addr_size: u8 = addr.ip_size()?.try_into().map_err(|_| {
+                io::Error::new(io::ErrorKind::InvalidInput, "failed to parse ip size")
+            })?;
+
+            w.put_u8(addr_size)?;
+            addr.scorex_serialize(w)?;
+
+            Ok(())
+        })?;
+
+        // Can't use Vec<ScorexSerializable> becuase we need the size as u8
```

Comment:
I'm actually not sure if this is the case, as it seems the underlying `put_u` functions cast everything to 64 sized ints

Reply:
Had a bit more of a look in the Scala codebase and it looks like that's part of the VLQ encoding

Reply:
Although you're right that `put_u32` for 0..256 values will be equivalent to `put_u8` I'd rather use `put_u8` to serialize elements manually. In the protocol, it's a one-byte so let's explicitly encode it in the type system. Actually, let's go one step further and make "illegal states unrepresentable" :). What do you think of making `PeerSpec::features` be of type `Option<BoundedVec<1, 255, PeerFeature>>`?

Reply:
Yep sounds good, I was a bit loose with the serialization types because of the upcasting so I'll go back through and make sure they're updated to the correct sizes 👍